lib/dirac_coarse.cpp

```diff
@@ -190,21 +193,72 @@ namespace quda {
+    if (!gpu_setup) {
 
-    // save the intermediate tunecache after the UV and VUV tune
-    saveTuneCache();
+      dirac->createCoarseOp(*Y_h, *X_h, *transfer, kappa, mass, Mu(), MuFactor());
+      // save the intermediate tunecache after the UV and VUV tune
+      saveTuneCache();
+      if (getVerbosity() >= QUDA_VERBOSE) printfQuda("About to build the preconditioned coarse clover\n");
+
+      createYhat(gpu_setup);
+
+      if (getVerbosity() >= QUDA_VERBOSE) printfQuda("Finished building the preconditioned coarse clover\n");
+      if (getVerbosity() >= QUDA_VERBOSE) printfQuda("About to create the preconditioned coarse op\n");
+
+      calculateYhat(*Yhat_h, *Xinv_h, *Y_h, *X_h, use_mma);
+
+    } else {
+
+      if (use_mma && dirac->isCoarse()) {
+
+        constexpr QudaGaugeFieldOrder gOrder = QUDA_MILC_GAUGE_ORDER;
+
+        GaugeFieldParam Y_param(*Y_d);
+        GaugeFieldParam X_param(*X_d);
+
+        Y_param.order = gOrder;
+        X_param.order = gOrder;
+
+        GaugeField *Y_order = cudaGaugeField::Create(Y_param);
+        GaugeField *X_order = cudaGaugeField::Create(X_param);
+
+        dirac->createCoarseOp(*Y_order, *X_order, *transfer, kappa, mass, Mu(), MuFactor());
+
+        // save the intermediate tunecache after the UV and VUV tune
+        saveTuneCache();
 
-    if (getVerbosity() >= QUDA_VERBOSE) printfQuda("About to build the preconditioned coarse clover\n");
+        if (getVerbosity() >= QUDA_VERBOSE) printfQuda("About to build the preconditioned coarse clover\n");
 
-    createYhat(gpu_setup);
+        createYhat(gpu_setup);
 
-    if (getVerbosity() >= QUDA_VERBOSE) printfQuda("Finished building the preconditioned coarse clover\n");
-    if (getVerbosity() >= QUDA_VERBOSE) printfQuda("About to create the preconditioned coarse op\n");
+        if (getVerbosity() >= QUDA_VERBOSE) printfQuda("Finished building the preconditioned coarse clover\n");
+        if (getVerbosity() >= QUDA_VERBOSE) printfQuda("About to create the preconditioned coarse op\n");
 
-    if (gpu_setup) createPreconditionedCoarseOp(*Yhat_d,*Xinv_d,*Y_d,*X_d);
-    else createPreconditionedCoarseOp(*Yhat_h,*Xinv_h,*Y_h,*X_h);
+        calculateYhat(*Yhat_d, *Xinv_d, *Y_order, *X_order, use_mma);
+
+        Y_d->copy(*Y_order);
+        X_d->copy(*X_order);
+
+        delete Y_order;
+        delete X_order;
+
+      } else {
+
+        dirac->createCoarseOp(*Y_d, *X_d, *transfer, kappa, mass, Mu(), MuFactor());
+
+        // save the intermediate tunecache after the UV and VUV tune
+        saveTuneCache();
+
+        if (getVerbosity() >= QUDA_VERBOSE) printfQuda("About to build the preconditioned coarse clover\n");
+
+        createYhat(gpu_setup);
+
+        if (getVerbosity() >= QUDA_VERBOSE) printfQuda("Finished building the preconditioned coarse clover\n");
+        if (getVerbosity() >= QUDA_VERBOSE) printfQuda("About to create the preconditioned coarse op\n");
+
+        calculateYhat(*Yhat_d, *Xinv_d, *Y_d, *X_d, use_mma);
```

Comment:
Temporary fix: `use_mma` should be replaced with `false`. Not sure if this is the most robust solution, though, because `use_mma` should still apply to `calculateYhat` even if we're not coarsening a coarse op?

Reply:
Update, this seems to only fix level 0 -> level 1

Reply:
This initial incident has now been fixed. Marking this conversation resolved, will create a new one as necessary.